pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp

```diff
@@ -0,0 +1,565 @@
+  auto loc = op.getLoc();
+  if (!inputTy) {
+    op.emitError("input should be ranked tensor type.");
+  }
+  auto definingOp = op.self().getDefiningOp();
+  auto shape = definingOp->getOperand(0);
+  SmallVector<Value, 4> dimSizes;
+  getListConstructElements(shape, dimSizes);
+  std::for_each(dimSizes.begin(), dimSizes.end(), [&](Value& dSize) {
+    dSize = rewriter.create<ToI64Op>(loc, dSize).getResult();
+    // dimSize: cast i64 -> i32
+    dSize = rewriter.create<arith::TruncIOp>(loc, rewriter.getI32Type(), dSize);
+    return dSize;
+  });
+
+  auto mhloShape =
+      rewriter.create<tensor::FromElementsOp>(op.getLoc(), dimSizes);
+
+  double fromDoubleValue, toDoubleValue;
+  if (!matchPattern(op.from(), m_TorchConstantFloat(&fromDoubleValue))) {
+    op.emitError("operand #1 should be scalar");
+  }
+  if (!matchPattern(op.to(), m_TorchConstantFloat(&toDoubleValue))) {
+    op.emitError("operand #2 should be scalar");
+  }
+  Value fromTensor = rewriter.create<mhlo::ConstantOp>(
+      op.getLoc(),
+      rewriter.getFloatAttr(inputTy.getElementType(), fromDoubleValue));
+  Value toTensor = rewriter.create<mhlo::ConstantOp>(
+      op.getLoc(),
+      rewriter.getFloatAttr(inputTy.getElementType(), toDoubleValue));
+
+  auto outType = getTypeConverter()
+                     ->convertType(op.getType())
+                     .template dyn_cast<TensorType>();
+  rewriter.replaceOpWithNewOp<mhlo::RngOp>(
+      op,
+      inputTy,
+      fromTensor,
+      toTensor,
+      mhloShape,
+      mhlo::RngDistribution::UNIFORM);
+  return success();
+}
+
+// Convert a Aten::LeakyRelu to HLO
+// LeakyRelu(x) = max(0, x) + negative_slop * min(0, x)
+template <>
+LogicalResult ConvertAtenOp<AtenLeakyReluOp>::matchAndRewrite(
+    AtenLeakyReluOp op,
+    OpAdaptor adaptor,
+    ConversionPatternRewriter& rewriter) const {
+  Location loc = op.getLoc();
+  Value input = adaptor.self();
+  Value negativeSlope = op.negative_slope();
+  auto inputTy = input.getType().template dyn_cast<RankedTensorType>();
+  if (!inputTy) {
+    return op.emitError(
+        "only RankedTensorType is supported in Aten LeakyReLU op.");
+  }
+
+  double scaleValue;
+  if (!matchPattern(negativeSlope, m_TorchConstantFloat(&scaleValue)))
+    return op->emitError(
+        "currently only scalar constants are supported for "
+        "negative_slope in MHLO operation");
+
+  Value zeroVal = chlo::getConstantLike(rewriter, loc, 0.0, input);
+  Value scaleVal = chlo::getConstantLike(rewriter, loc, scaleValue, input);
+
+  Value leakyActivationVal = rewriter.create<mhlo::MulOp>(
+      loc, getTypeConverter()->convertType(op.getType()), input, scaleVal);
+
+  Value compareGtZero = rewriter.create<mhlo::CompareOp>(
+      loc, input, zeroVal, mhlo::ComparisonDirection::GT);
+
+  Value output = rewriter.create<mhlo::SelectOp>(
+      loc, inputTy, compareGtZero, input, leakyActivationVal);
+  rewriter.replaceOpWithNewOp<mhlo::ConvertOp>(
```

Comment:
I see `ConvertOp` is intensively used, why it's always added at the end?

Reply:
The output tensor dtype and rank cannot change after the op conversion. But this might be true in a dynamic shape scenario because it relies on shape inference, but our shape inference is not complete for now.

It's introduced to run out of type checks before and after op conversion since `mhlo::ConvertOp` tolerates different input and output types. I think we should not have the `ConvertOp`.

Reply:
I had reduce the `ConvertOp` usages in this PR